cassiopeia/simulator/UniformLeafSubsampler.py

```diff
@@ -0,0 +1,72 @@
+import abc
+import networkx as nx
+import numpy as np
+from typing import Optional
+
+from cassiopeia.data import CassiopeiaTree
+from cassiopeia.simulator.LeafSubsampler import LeafSubsampler, LeafSubsamplerError
+
+
+class UniformLeafSubsampler(LeafSubsampler):
+    def __init__(
+        self, ratio: Optional[float] = None, n_leaves: Optional[int] = None
+    ):
+        """
+        Uniformly subsample leaf samples.
+
+        If 'ratio' is provided, samples 'ratio' of the leaves, rounded down,
+        uniformly at random. If instead 'n_leaves' is provided, 'n_leaves' of
+        the leaves are sampled uniformly at random. Only one of the two
+        criteria can be provided.
+
+        Args:
+            ratio: Specifies the number of leaves to be sampled as a ratio of
+                the total number of leaves
+            n_leaves: Explicitly specifies the number of leaves to be sampled
+        """
+        if ratio is None and n_leaves is None:
+            raise LeafSubsamplerError(
+                "At least one of 'ratio' and 'n_leaves' " "must be specified."
+            )
+        if ratio is not None and n_leaves is not None:
+            raise LeafSubsamplerError(
+                "Exactly one of 'ratio' and 'n_leaves'" "must be specified."
+            )
+        self.__ratio = ratio
+        self.__n_leaves = n_leaves
```

Comment:
I based this off of what Sebastian was doing. I think the belief is that once we establish the `ratio` and `n_leaves` inputs, we don't want the user to be able to go in and alter those. I think this is just for consistency and to avoid confusion, since if the user wanted to specify different `ratio` and `n_leaves`  they would easily create a new subsampler object. Additionally, we're checking to see if both `ratio` and `n_leaves` are specified in the constructor (if they are, we error), and we don't want the user to initialize the object and then add either of these afterwards